File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        
        answerTree = TreeNode()
        currentNode = answerTree
        
        
        
        def dfs(root):
            if not root:
                return None
            
            dfs(root.left)
            nonlocal currentNode
            root.left= None
            currentNode.right = root
            currentNode = currentNode.right
            
            dfs(root.right)
            
        dfs(root)
        return answerTree.right
        

File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* newRoot = new TreeNode;
    TreeNode* iter = newRoot;
    
    TreeNode* increasingBST(TreeNode* root) {
        
        if (!root)
            return nullptr;
        
        
        increasingBST(root->left);
        iter->right = new TreeNode(root->val);
        
        iter = iter->right;
        
        increasingBST(root->right);
        
        
        
        return newRoot->right;
        
        
        
        
    }
};